### PR TITLE
TimePicker

### PR DESCRIPTION
**Pull request by MargaritaChe**
### Description of changes ###

Added TimePicker control. Also added the TimePicker popup as part of the DateTimePicker control.

### API Changes ###
 
No one

### Affected Platforms ###

- WPF

### Behavioral changes ###

Changing the time in the DateTimePicker is now possible using a pop-up window containing the clock face.

### CODE REVIEW

File: src/Orc.Controls.Example/Orc.Controls.Example.csproj [hunk 1]
```diff
@@ -14,12 +14,20 @@
     <NoWarn>$(NoWarn);SA1652</NoWarn>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (GeertvanHorrik) [inline src/Orc.Controls.Example/Orc.Controls.Example.csproj:17]: This can be removed

File: src/Orc.Controls.Example/Orc.Controls.Example.csproj [hunk 2]
```diff
@@ -14,12 +14,20 @@
...
+    <Compile Remove="ViewModels\TimePicker.cs" />
+  </ItemGroup>
+
   <ItemGroup>
     <PackageReference Include="Catel.Fody" Version="4.4.4" PrivateAssets="all" />
     <PackageReference Include="Fody" Version="6.2.0" PrivateAssets="all">
       <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
     </PackageReference>
     <PackageReference Include="LoadAssembliesOnStartup.Fody" Version="4.5.0" PrivateAssets="all" />
+    <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0">
```
Comment (GeertvanHorrik) [inline src/Orc.Controls.Example/Orc.Controls.Example.csproj:27]: Why is this one needed?

File: src/Orc.Controls.Example/ViewModels/TimePickerViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        public TimeSpan? Time { get; set; }
+        public string TimeValueString { get; set; }
+        public Meridiem AmPm { get; set; }
+        public double HourThickness { get; set; }
+        public double MinuteThickness { get; set; }
+        public double HourTickThickness { get; set; }
+        public double MinuteTickThickness { get; set; }
+        public double ClockBorderThickness { get; set; }
+        public Command SetNull { get; }
```
Comment (GeertvanHorrik) [inline src/Orc.Controls.Example/ViewModels/TimePickerViewModel.cs:41]: Please keep commands organized (see other examples)

File: src/Orc.Controls.Example/Views/DateTimePicker.xaml [hunk 4]
```diff
@@ -8,8 +8,7 @@
     <UserControl.Resources>
     </UserControl.Resources>
 
-    <!-- Content -->
-    <StackPanel>
+    <!-- Content -->    <StackPanel>
```
Comment (GeertvanHorrik) [inline src/Orc.Controls.Example/Views/DateTimePicker.xaml:11]: Unnecessary change

File: src/Orc.Controls.Example/Views/FilterBox.xaml.cs [hunk 5]
```diff
@@ -17,9 +18,9 @@ public FilterBox()
             InitializeComponent();
         }
 
-        private void OnFilterBoxControlInitializingAutoCompletionService(object sender, Controls.InitializingAutoCompletionServiceEventArgs e)
+        private void OnFilterBoxControlInitializingAutoCompletionService(object sender, Orc.Controls.InitializingAutoCompletionServiceEventArgs e)
```
Comment (GeertvanHorrik) [inline src/Orc.Controls.Example/Views/FilterBox.xaml.cs:21]: Unnecessary change

File: src/Orc.Controls.Example/Views/TimePicker.xaml [hunk 6]
```diff
@@ -0,0 +1,57 @@
+﻿<catel:UserControl x:Class="Orc.Controls.Example.Views.TimePicker"
+                   xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+                   xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+				   xmlns:catel="http://schemas.catelproject.com"
+				   xmlns:tp="clr-namespace:Orc.Controls;assembly=Orc.Controls">
+
+    <!-- Resources -->
```
Comment (GeertvanHorrik) [inline src/Orc.Controls.Example/Views/TimePicker.xaml:7]: Can be removed (no resources used)

File: src/Orc.Controls.Example/Views/TimePicker.xaml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    <UserControl.Resources>
+    </UserControl.Resources>
+
+    <!-- Content -->
+    <StackPanel>
+        <Grid>
```
Comment (GeertvanHorrik) [inline src/Orc.Controls.Example/Views/TimePicker.xaml:13]: Please use StackGrid if applicable

File: src/Orc.Controls.Example/Views/TimePicker.xaml.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Orc.Controls.Example.Views
+{
+    using System;
+    using System.ComponentModel;
+    using System.Windows;
+    using Catel.Data;
+    using Enums;
+
+    public partial class TimePicker: INotifyPropertyChanged
```
Comment (GeertvanHorrik) [inline src/Orc.Controls.Example/Views/TimePicker.xaml.cs:17]: : INotifyPropertyChanged is not needed

File: src/Orc.Controls.Tests/Orc.Controls.Tests.csproj [hunk 9]
```diff
@@ -21,6 +21,10 @@
     <PackageReference Include="LoadAssembliesOnStartup.Fody" Version="4.5.0" PrivateAssets="all" />
     <PackageReference Include="MethodTimer.Fody" Version="3.1.2" PrivateAssets="all" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.6.1" />
+    <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0">
```
Comment (GeertvanHorrik) [inline src/Orc.Controls.Tests/Orc.Controls.Tests.csproj:24]: Why this reference?

File: src/Orc.Controls/Helpers/DateTimeFormatHelper.cs [hunk 10]
```diff
@@ -64,7 +64,7 @@ public static string[] Split(string format, char[] formatCharacters)
             return parts.ToArray();
         }
 
-        public static DateTimeFormatInfo GetDateTimeFormatInfo(string format, bool isDateOnly = false)
+        public static DateTimeFormatInfo GetDateTimeFormatInfo(string format, bool isDateRequired, bool isDateOnly = false)
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Helpers/DateTimeFormatHelper.cs:67]: Seems like a breaking change. Let's discuss the proper way forward to introduce this.

File: src/Orc.Controls/Orc.Controls.csproj [hunk 11]
```diff
@@ -57,6 +61,12 @@
     <Folder Include="Controls\EditableControls\Behaviors\Base\" />
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Orc.Controls.csproj:64]: Please remove

File: src/Orc.Controls/Orc.Controls.csproj [hunk 12]
```diff
@@ -34,6 +34,10 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
     <PackageReference Include="MethodTimer.Fody" Version="3.1.2" PrivateAssets="all" />
+    <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0">
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Orc.Controls.csproj:37]: Please remove

**Review by GeertvanHorrik — CHANGES_REQUESTED** [review]
Reviewed most changes, please make the necessary changes and we'll review again.

File: src/Orc.Controls/Helpers/DateTimeFormatHelper.cs (hunk 10 above)
Comment (MargaritaChe) [reply src/Orc.Controls/Helpers/DateTimeFormatHelper.cs]: These changes were unnecessary. I canceled them.

File: src/Orc.Controls.Example/Views/TimePicker.xaml (hunk 7 above)
Comment (mkhomutov) [reply src/Orc.Controls.Example/Views/TimePicker.xaml:9]: Why Grid wrapped by StackGrid?
Just replace Grid with StackGrid

File: src/Orc.Controls/Controls/DateTimePicker/DateTimePicker.cs [hunk 13]
```diff
@@ -608,6 +662,50 @@ private void UnsubscribeFromCalendarEvents()
...
+        }
+        private void OnTimeValueChanged(TimeSpan? newTimeValue)
+        {
+            if (newTimeValue != null)
+            {
+                DateTime newDateTime;
+                if (HideTime == true)
+                {
+                    SetCurrentValue(HideTimeProperty, false);
+                }
```
Comment (mkhomutov) [inline src/Orc.Controls/Controls/DateTimePicker/DateTimePicker.cs:688]: the new line between if statements

File: src/Orc.Controls/Controls/DateTimePicker/DateTimePicker.cs [hunk 14]
```diff
@@ -608,6 +662,50 @@ private void UnsubscribeFromCalendarEvents()
...
+                if(Value != null)
+                {
+                    newDateTime = Value.Value;
+                }
+                else
+                {
+                    newDateTime = DateTime.Now;
+                }
```
Comment (mkhomutov) [inline src/Orc.Controls/Controls/DateTimePicker/DateTimePicker.cs:696]: add empty line

File: src/Orc.Controls/Controls/DateTimePicker/DateTimePicker.cs [hunk 15]
```diff
@@ -1174,6 +1272,40 @@ private void OnAmPmListTextBoxValueChanged(object sender, EventArgs e)
...
+            if (newValue == Meridiem.AM)
+            {
+                if (newDateTime.Hour >= 12)
+                {
+                    diffTime = newDateTime.TimeOfDay.Subtract(hours);
+                    var diffDate = new DateTime(newDateTime.Year, newDateTime.Month, newDateTime.Day, newDateTime.Hour - 12, newDateTime.Minute, newDateTime.Second);
+                    SetCurrentValue(ValueProperty, diffDate);
+                }
+            }
+            else
```
Comment (mkhomutov) [inline src/Orc.Controls/Controls/DateTimePicker/DateTimePicker.cs:1298]: try to avoid `else if`

**Review by mkhomutov — CHANGES_REQUESTED** [review]
looks good. just a few small fixes required.

**Comment by ceo1647** [comment]
it doesn't support dark mode in popup
image

**Comment by ceo1647** [comment]
as you can see width1 == width2 
image
but here...
image
also maybe it's better to draw background and border for Clock popup

**Comment by ceo1647** [comment]
we should Capture mouse cursor when dragging and dropping clock hand... now if we go out the clock border the dragging process stops

**Comment by ceo1647** [comment]
i think it would be convenient if we have an ability to hide/show labels for hours?

**Comment by GeertvanHorrik** [comment]
Please add link to original repo as requested by the original developer @roy-t

Add it to LICENSE (similar to how we deal with Catel, keep alphabetically ordered):

https://github.com/roy-t/TimePicker

**Comment by ceo1647** [comment]
layout for clock surface is very good... button layout should be fixed
image

**Comment by ceo1647** [comment]
just a thought... maybe we need to handle MouseOver and MousePressed events on clock hands...also think how it will look in DisabledMode

**Review by mkhomutov — CHANGES_REQUESTED** [review]
image

AM/PM button is too small

**Comment by mkhomutov** [comment]
Thanks, looks good now
